packages/rendermime/src/renderers.ts

```diff
@@ -554,15 +554,100 @@ function splitShallowNode<T extends Node>(
   at: number
 ): { pre: T; post: T } {
   const pre = node.cloneNode() as T;
-  pre.textContent = node.textContent?.substr(0, at) as string;
+  pre.textContent = node.textContent?.slice(0, at) as string;
   const post = node.cloneNode() as T;
-  post.textContent = node.textContent?.substr(at) as string;
+  post.textContent = node.textContent?.slice(at) as string;
   return {
-    pre: pre,
-    post: post
+    pre,
+    post
   };
 }
 
+
+/**
+ * Iterate over some nodes, while tracking cumulative start and end position.
+ */
+function* nodeIter<T extends Node>(nodes: T[]): IterableIterator<{node: T, start: number, end: number, isText: boolean}> {
+  let start = 0;
+  let end;
+  for (let node of nodes) {
+    end = start + (node.textContent?.length || 0);
+    yield {
+      node,
+      start,
+      end,
+      isText: node.nodeType === Node.TEXT_NODE,
+    };
+    start = end;
+  }
+}
+
+
+/**
+ * Align two collections of nodes.
+ *
+ * If a text node in one collections spans an element in the other, yield the spanned elements.
+ * Otherwise, split the nodes such that yielded pair start and stop on the same position.
+ */
+function* alignedNodes<T extends Node, U extends Node>(a: T[], b: U[]): IterableIterator<[T, null] | [null, U] | [T, U]> {
+  let iterA = nodeIter(a);
+  let iterB = nodeIter(b);
+  let nA = iterA.next();
+  let nB = iterB.next();
+  while (!nA.done && !nB.done) {
+    let A = nA.value;
+    let B = nB.value;
+
+    if (A.isText && A.start <= B.start && A.end >= B.end) {
```

Comment:
This is neat, thank you! I think it would be great to also add a test case for each of the conditions in `packages/rendermime/test/factories.spec.ts`.

Reply:
Thanks, I added enough tests to get the coverage to 100%. It unveiled a log error! 👍